### ISSUE TITLE
Server side token

### ISSUE THREAD

**Issue by vauvenal5** [nextcloud/Android-SingleSignOn#297]
Is it possible to use the Android SSO flow strictly for authentication or is it necessary to run all network communication over the Android files client?

**Comment by stefan-niedermann** [comment]
Yes, Deck and Notes app do not evem have a own network stack. All communication runs through the SSO.

**Comment by David-Development** [comment]
The idea behind this SSO library is that the authentication details never leave the official nextcloud app. Therefore all communication to the server need to go through the Android Files Client. Do you have any concern why you wouldn't want to do that? If you like to use retrofit, we have a compatible API for that so you won't really notice much of a difference - except that the login is much easier and you don't need to worry about difficult server setups.

**Comment by vauvenal5** [comment]
Well my app is Flutter app, so written in Dart. I guess this kind of limits the possibilities of integration. I suppose I could access this API your provide retrofit with any arbitrary HTTP client? Does reassemble the Nextcloud server API? I must have overseen it in the documentation.

**Comment by tobiasKaminsky** [comment]
The server API you have to call / maintain yourself.
SSO is only handling network traffic for you.

**Comment by vauvenal5** [comment]
Sorry, let me rephrase that, if I understood it correctly, for every request the SSO library contacts the Files-App and the Files-App does the actual network request. How does to SSO library communicate with the Files-App and is this endpoint at the Files-App somewhere documented?

To make this work with Flutter I either have to write a library that does basically the same as this one and communicates directly with the Files-App instead of the Nextcloud server. Or, I have to write a wrapper library that bridges between Java and Flutter and exposes the functionality provided by the SSO library. Just trying to understand what might be the cheaper approach here.

**Comment by tobiasKaminsky** [comment]
https://github.com/nextcloud/Android-SingleSignOn#flow-diagram might help
else handling requests is in InputStreamBinder on Files app.

I fear that rebuilding this SSO for Flutter is very time consuming and if there are not many Flutter apps right now, it might be not worth it…

**Comment by David-Development** [comment]
>  if I understood it correctly, for every request the SSO library contacts the Files-App and the Files-App does the actual network request.

Yes, that is correct.

> How does to SSO library communicate with the Files-App

It uses IPC (Interprocess Communication using AIDL (Android Interface Definition Language)

> and is this endpoint at the Files-App somewhere documented?

not really - you can find the AIDL [here](https://github.com/nextcloud/android/blob/master/src/main/aidl/com/nextcloud/android/sso/aidl/IInputStreamService.aidl). But as Tobias said, I think rebuilding this in Flutter will be quite some work as there are some caveats to implementing it as for example you need to bind a service to the files app in order to use the IPC and then you need to support the files app + files app beta version and you need to handle the unbinding of the services etc. etc. etc. 

I haven't really used flutter much yet so I can't say much about it but I'd probably try to integrate this library somehow if possible rather than rebuilding this library from scratch again - I mean we can give you some pointers and you have a few working examples in Java / Kotlin but I'd still be quite time consuming.

**Comment by vauvenal5** [comment]
Yes, I assumed as much. To write an entire wrapping might proof to be just as much work. So I will probably leave this be for now and wait to see how the Nextcloud Flutter landscape develops.

Thx so far, I believe this can be closed.

**Comment by summersab** [comment]
I'm going to chime in. It looks like there are a number of Flutter apps popping up, actually. The two I am most interested in using are NC Passwords and Nextcloud Yaga (gallery):
https://gitlab.com/joleaf/nc-passwords-app
https://github.com/vauvenal5/yaga

NC Passwords looks pretty promising, and it includes autofill support on Android. They have an open issue regarding SSO, so it might be worth contacting the developer:
https://gitlab.com/joleaf/nc-passwords-app/-/issues/28

**Comment by sepo83** [comment]
Let me add another very good nextcloud android app written in dart/flutter:
https://github.com/Teifun2/nextcloud-cookbook-flutter

I would really appreciate if SSO would work for all those apps. Makes things much easier and first of all more secure for the user.

**Comment by provokateurin** [comment]
https://github.com/provokateurin/nextcloud-neon/issues/65 for anyone interested in discussing how to implement this in flutter :)
